modules/reindex/src/main/java/org/elasticsearch/index/reindex/AbstractAsyncBulkIndexByScrollAction.java

```diff
@@ -86,6 +86,9 @@ protected BulkRequest buildBulk(Iterable<SearchHit> docs) {
         Map<String, Object> scriptCtx = null;
 
         for (SearchHit doc : docs) {
+            if (doc.isSourceEmpty()) {
+                throw new IllegalArgumentException("[" + doc.index() + "][" + doc.type() + "][" + doc.id() + "] didn't store _source");
```

Comment:
I can certainly check the mapping but I imagined that a document that made it into the index without any source always did so because the source was disabled. I'll play around and see if I can get that to happen without source being disabled.

I guess if it can happen then I can use an empty JSON structure as the source.